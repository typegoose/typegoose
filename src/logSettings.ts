import * as logger from 'loglevel';

export { logger };

export const setLogLevel = logger.setLevel;
export const LogLevels = logger.levels;
logger.setDefaultLevel(LogLevels.WARN);
