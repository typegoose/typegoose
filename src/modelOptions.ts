import { DecoratorKeys } from './internal/constants';
import { assignGlobalModelOptions, assignMetadata } from './internal/utils';
import type { IModelOptions } from './types';

/**
 * Define Options for the Class
 * @param options Options
 * @example Example:
 * ```ts
 * @modelOptions({ schemaOptions: { timestamps: true } })
 * class ClassName {}
 *
 * // The default Class "TimeStamps" can be used for type information and options already set
 * ```
 */
export function modelOptions(options: IModelOptions): ClassDecorator {
  return (target: any) => {
    assignGlobalModelOptions(target);
    assignMetadata(DecoratorKeys.ModelOptions, options, target);
  };
}

// Export it PascalCased
export { modelOptions as ModelOptions };
