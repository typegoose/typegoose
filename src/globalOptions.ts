import { Severity } from './internal/constants';
import { globalOptions } from './internal/data';
import { assertion } from './internal/utils';
import { logger } from './logSettings';
import type { IGlobalOptions } from './types';

/**
 * Set Typegoose's global Options
 */
export function setGlobalOptions(options: IGlobalOptions) {
  assertion(typeof options === 'object', new TypeError('"options" argument needs to be an object!'));

  logger.info('"setGlobalOptions" got called with', options);

  for (const key of Object.keys(options)) {
    globalOptions[key] = Object.assign({}, globalOptions[key], options[key]);
  }

  logger.info('new Global Options:', options);
}

/**
 * Parse Typegoose Environment Variables and apply them
 */
export function parseENV(): void {
  logger.info('"parseENV" got called');

  const options: IGlobalOptions = {
    globalOptions: {},
    options: {
      allowMixed:
        process.env.TG_ALLOW_MIXED && process.env.TG_ALLOW_MIXED in Severity
          ? mapValueToSeverity(process.env.TG_ALLOW_MIXED)
          : globalOptions.options?.allowMixed,
    },
  };

  setGlobalOptions(options);
}

/**
 * Maps strings to the number
 * -> This function is specifically build for "Severity"-Enum
 * @throws {Error} if not in range of the "Severity"-Enum
 * @example
 * ```ts
 * mapValueToSeverity("WARN") === 1
 * mapValueToSeverity("1") === 1
 * // now internal use
 * mapValueToSeverity(1) === 1
 * ```
 * @param value The value to check for
 */
function mapValueToSeverity(value: string | number): Severity {
  assertion(value in Severity, new Error(`"value" is not in range of "Severity"! (got: ${value})`));

  if (typeof value === 'number') {
    return value;
  }

  return mapValueToSeverity(Severity[value]);
}
