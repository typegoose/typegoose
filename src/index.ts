import { DecoratorKeys } from './internal/constants';
import { getName } from './internal/utils';
import { logger } from './logSettings';
import type { BeAnObject, IIndexArray, IndexOptions } from './types';

/**
 * Defines an index (most likely compound) for this schema.
 * @param fields Which fields to give the Options
 * @param options Options to pass to MongoDB driver's createIndex() function
 * @example Example:
 * ```ts
 * @index({ article: 1, user: 1 }, { unique: true })
 * class ClassName {}
 * ```
 */
export function index<T extends BeAnObject = BeAnObject>(
  fields: Partial<Record<keyof T, string | -1 | 1>>,
  options?: IndexOptions<T>
): ClassDecorator {
  return (target: any) => {
    logger.info('Adding "%o" Indexes to %s', { fields, options }, getName(target));
    const indices: IIndexArray<any>[] = Array.from(Reflect.getMetadata(DecoratorKeys.Index, target) ?? []);
    indices.push({ fields, options });
    Reflect.defineMetadata(DecoratorKeys.Index, indices, target);
  };
}

// Export it PascalCased
export { index as Index };
