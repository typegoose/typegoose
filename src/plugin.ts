import { DecoratorKeys } from './internal/constants';
import { getName } from './internal/utils';
import { logger } from './logSettings';
import type { Func, IPluginsArray } from './types';

/**
 * Add a Middleware-Plugin
 * @param mongoosePlugin The Plugin to plug-in
 * @param options Options for the Plugin, if any
 * @example Example:
 * ```ts
 * @plugin(findOrCreate)
 * class ClassName {}
 * ```
 */
export function plugin<TFunc extends Func, TParams = Parameters<TFunc>[1]>(mongoosePlugin: TFunc, options?: TParams): ClassDecorator {
  // don't check if options is an object, because any plugin could make it anything
  return (target: any) => {
    logger.info('Adding plugin "%s" to "%s" with options: "%o"', mongoosePlugin?.name ?? '<anonymous>', getName(target), options);
    const plugins: IPluginsArray<any>[] = Array.from(Reflect.getMetadata(DecoratorKeys.Plugins, target) ?? []);
    plugins.push({ mongoosePlugin, options });
    Reflect.defineMetadata(DecoratorKeys.Plugins, plugins, target);
  };
}

// Export it PascalCased
export { plugin as Plugins };
