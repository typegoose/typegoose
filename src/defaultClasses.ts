import type { Types } from 'mongoose';
import { modelOptions } from './modelOptions';
import type { AnyParamConstructor, DocumentType, RefType } from './types';

@modelOptions({ schemaOptions: { timestamps: true } })
/**
 * This class can be used for already existing type information for the Timestamps
 */
export abstract class TimeStamps {
  public createdAt?: Date;
  public updatedAt?: Date;
}

/**
 * This Interface can be used when "_id" and "id" need to be defined in types
 */
export interface Base<IDType extends RefType = Types.ObjectId> {
  _id: IDType;
  /**
   * This getter/setter dosnt exist if "schemaOptions.id" being set to "false"
   */
  id: string;
}

export interface FindOrCreateResult<T> {
  created: boolean;
  doc: DocumentType<T>;
}

/**
 * This class contains all types for the module "mongoose-findorcreate"
 */
export abstract class FindOrCreate {
  public static findOrCreate: <T extends FindOrCreate>(
    this: AnyParamConstructor<T>,
    condition: any,
    createWith?: any
  ) => Promise<FindOrCreateResult<T>>;
}
