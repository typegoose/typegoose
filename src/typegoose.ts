/* imports */
import * as mongoose from 'mongoose';
import 'reflect-metadata';
import * as semver from 'semver';
import {
  assertion,
  assertionIsClass,
  getMergedModelOptions,
  getName,
  isCachingEnabled,
  isGlobalCachingEnabled,
  isNullOrUndefined,
  mapModelOptionsToNaming,
  warnNotMatchingExisting,
} from './internal/utils';

// using "typeof process", because somehow js gives a ReferenceError when using "process === undefined" in browser
/* istanbul ignore next */
if (typeof process !== 'undefined' && !isNullOrUndefined(process?.version) && !isNullOrUndefined(mongoose?.version)) {
  // for usage on client side
  /* istanbul ignore next */
  if (semver.lt(mongoose?.version, '8.1.0')) {
    throw new Error(`Please use mongoose 8.1.0 or higher (Current mongoose: ${mongoose.version}) [E001]`);
  }

  /* istanbul ignore next */
  if (semver.lt(process.version.slice(1), '14.17.0')) {
    throw new Error('You are using a NodeJS Version below 14.17.0, Please Upgrade! [E002]');
  }
}

import { parseENV, setGlobalOptions } from './globalOptions';
import { DecoratorKeys } from './internal/constants';
import { constructors, models } from './internal/data';
import { _buildSchema } from './internal/schema';
import { logger } from './logSettings';
import { isModel } from './typeguards';
import type {
  AnyParamConstructor,
  BeAnObject,
  DocumentType,
  IModelOptions,
  Ref,
  ReturnModelType,
  SubDocumentType,
  ArraySubDocumentType,
  IBuildSchemaOptions,
} from './types';
import { CacheDisabledError, ExpectedTypeError, FunctionCalledMoreThanSupportedError, NotValidModelError } from './internal/errors';

/* exports */
// export the internally used "mongoose", to not need to always import it
export { mongoose, setGlobalOptions };
export { setLogLevel, LogLevels } from './logSettings';
export * from './prop';
export * from './hooks';
export * from './plugin';
export * from './indexes';
export * from './modelOptions';
export * from './queryMethod';
export * from './typeguards';
export * as defaultClasses from './defaultClasses';
export * as errors from './internal/errors';
export * as types from './types';
// the following types are re-exported (instead of just in "types") because they are often used types
export { DocumentType, Ref, ReturnModelType, SubDocumentType, ArraySubDocumentType };
export { getClass, getName } from './internal/utils';
export { Severity, PropType } from './internal/constants';

parseENV(); // call this before anything to ensure they are applied

/**
 * Build a Model From a Class
 * @param cl The Class to build a Model from
 * @param options Overwrite Options, like for naming or general SchemaOptions the class gets compiled with
 * @returns The finished Model
 * @public
 * @example
 * ```ts
 * class ClassName {}
 *
 * const NameModel = getModelForClass(ClassName);
 * ```
 */
export function getModelForClass<U extends AnyParamConstructor<any>, QueryHelpers = BeAnObject>(cl: U, options?: IModelOptions) {
  assertionIsClass(cl);
  const rawOptions = typeof options === 'object' ? options : {};
  const overwriteNaming = mapModelOptionsToNaming(rawOptions); // use "rawOptions" instead of "mergedOptions" to consistently differentiate between classes & models

  const mergedOptions = getMergedModelOptions(rawOptions, cl);
  const name = getName(cl, overwriteNaming);

  if (isCachingEnabled(mergedOptions.options?.disableCaching) && models.has(name)) {
    return models.get(name) as ReturnModelType<U, QueryHelpers>;
  }

  const modelFn =
    mergedOptions?.existingConnection?.model.bind(mergedOptions.existingConnection) ??
    mergedOptions?.existingMongoose?.model.bind(mergedOptions.existingMongoose) ??
    mongoose.model.bind(mongoose);

  const compiledModel: mongoose.Model<any> = modelFn(name, buildSchema(cl, mergedOptions));

  return addModelToTypegoose<U, QueryHelpers>(compiledModel, cl, {
    existingMongoose: mergedOptions?.existingMongoose,
    existingConnection: mergedOptions?.existingConnection,
    disableCaching: mergedOptions.options?.disableCaching,
  });
}

/**
 * Get Model from internal cache
 * @param key Model's name key
 * @example
 * ```ts
 * class ClassName {}
 * getModelForClass(ClassName); // build the model
 * const NameModel = getModelWithString<typeof ClassName>("ClassName");
 * ```
 */
export function getModelWithString<U extends AnyParamConstructor<any>, QueryHelpers = BeAnObject>(
  key: string
): undefined | ReturnModelType<U, QueryHelpers> {
  assertion(typeof key === 'string', () => new ExpectedTypeError('key', 'string', key));
  assertion(isGlobalCachingEnabled(), () => new CacheDisabledError('getModelWithString'));

  return models.get(key) as any;
}

/**
 * Generates a Mongoose schema out of class props, iterating through all parents
 * @param cl The Class to build a Schema from
 * @param options Overwrite Options, like for naming or general SchemaOptions the class gets compiled with
 * @returns Returns the Build Schema
 * @example
 * ```ts
 * class ClassName {}
 * const NameSchema = buildSchema(ClassName);
 * const NameModel = mongoose.model("Name", NameSchema);
 * ```
 */
export function buildSchema<U extends AnyParamConstructor<any>>(
  cl: U,
  options?: IModelOptions
): mongoose.Schema<DocumentType<InstanceType<U>>> {
  assertionIsClass(cl);

  const overwriteNaming = mapModelOptionsToNaming(options);
  logger.debug('buildSchema called for "%s"', getName(cl, overwriteNaming));

  // dont re-run the merging if already done so before (like in getModelForClass)
  const mergedOptions = getMergedModelOptions(options, cl);

  let sch: mongoose.Schema<DocumentType<InstanceType<U>>> | undefined = undefined;
  /** Parent Constructor */
  let parentCtor = Object.getPrototypeOf(cl.prototype).constructor;
  /* This array is to execute from lowest class to highest (when extending) */
  const parentClasses: [AnyParamConstructor<any>, IBuildSchemaOptions][] = [];
  /** Options for the next lower class that gets unshifted to {@link parentClasses} (ie the super-class) */
  let superOptions: IBuildSchemaOptions = {};

  // first run for some options based on input class options (if any), because the while-loop is for the prototypes (if any)
  {
    // get new options because "mergedOptions" is merged with lower options, but "upperOptions" requires the "own" version, if any
    const mergedOwnOptions = getMergedModelOptions(options, cl, true);
    applySuperOptions(superOptions, mergedOwnOptions);
  }

  // iterate trough all parents to the lowest class
  while (parentCtor?.name !== 'Object') {
    // add lower classes (when extending) to the front of the array to be processed first
    parentClasses.unshift([parentCtor, superOptions]);

    // clone object, because otherwise it will affect the upper classes too because the same reference is used
    superOptions = { ...superOptions };

    {
      // only get the own metadata, possible because "upperOptions" at the moment only requires the "own" metadata
      const ropt: IModelOptions = Reflect.getOwnMetadata(DecoratorKeys.ModelOptions, parentCtor) ?? {};

      applySuperOptions(superOptions, ropt);
    }

    // set next parent
    parentCtor = Object.getPrototypeOf(parentCtor.prototype).constructor;
  }

  // iterate and build class schemas from lowest to highest (when extending classes, the lower class will get build first) see https://github.com/typegoose/typegoose/pull/243
  for (const [parentClass, extraOptions] of parentClasses) {
    // extend schema
    sch = _buildSchema(parentClass, sch!, mergedOptions, false, undefined, extraOptions);
  }

  // get schema of current model
  sch = _buildSchema(cl, sch, mergedOptions, true, overwriteNaming);

  return sch;
}

/**
 * Apply options to "superOptions" object, based on "modelOptions"
 * @param superOptions The "superOptions" object
 * @param modelOptions The Model Options of the current class
 */
function applySuperOptions(superOptions: IBuildSchemaOptions, modelOptions: IModelOptions) {
  // only affect options of lower classes, not the class the options are from
  if (modelOptions.options?.disableLowerIndexes) {
    superOptions.buildIndexes = false;
  }
}

/**
 * Add a Class-Model Pair to the Typegoose Cache
 * This can be used to add custom Models to Typegoose, with the type information of "cl"
 * Note: no guarrantee that the type information is fully correct when used manually
 * @param model The Model to store
 * @param cl The Class to store
 * @param options Overwrite existingMongoose or existingConnection
 * @example
 * ```ts
 * class ClassName {}
 *
 * const schema = buildSchema(ClassName);
 * // modifications to the schema can be done
 * const model = addModelToTypegoose(mongoose.model("Name", schema), ClassName);
 * ```
 */
export function addModelToTypegoose<U extends AnyParamConstructor<any>, QueryHelpers = BeAnObject>(
  model: mongoose.Model<any>,
  cl: U,
  options?: { existingMongoose?: mongoose.Mongoose; existingConnection?: any; disableCaching?: boolean }
) {
  const mongooseModel = options?.existingMongoose?.Model || options?.existingConnection?.base?.Model || mongoose.Model;

  assertion(model.prototype instanceof mongooseModel, new NotValidModelError(model, 'addModelToTypegoose.model'));
  assertionIsClass(cl);

  // only check cache after the above checks, just to make sure they run
  if (!isCachingEnabled(options?.disableCaching)) {
    logger.info('Caching is not enabled, skipping adding');

    return model as ReturnModelType<U, QueryHelpers>;
  }

  const name = model.modelName;

  assertion(
    !models.has(name),
    new FunctionCalledMoreThanSupportedError(
      'addModelToTypegoose',
      1,
      `This was caused because the model name "${name}" already exists in the typegoose-internal "models" cache`
    )
  );

  if (constructors.get(name)) {
    logger.info('Class "%s" already existed in the constructors Map', name);
  }

  models.set(name, model);
  constructors.set(name, cl);

  return models.get(name) as ReturnModelType<U, QueryHelpers>;
}

/**
 * Deletes a existing model so that it can be overwritten with another model
 * (deletes from mongoose.connection and typegoose models cache and typegoose constructors cache)
 * @param name The Model's mongoose name
 * @example
 * ```ts
 * class ClassName {}
 * const NameModel = getModelForClass(ClassName);
 * deleteModel("ClassName");
 * ```
 */
export function deleteModel(name: string) {
  assertion(typeof name === 'string', () => new ExpectedTypeError('name', 'string', name));
  assertion(isGlobalCachingEnabled(), () => new CacheDisabledError('deleteModelWithClass'));

  logger.debug('Deleting Model "%s"', name);

  const model = models.get(name);

  if (!isNullOrUndefined(model)) {
    model.db.deleteModel(name);
  }

  models.delete(name);
  constructors.delete(name);
}

/**
 * Delete a model, with the given class
 * Same as "deleteModel", only that it can be done with the class instead of the name
 * @param cl The Class to delete the model from
 * @example
 * ```ts
 * class ClassName {}
 * const NameModel = getModelForClass(ClassName);
 * deleteModelWithClass(ClassName);
 * ```
 */
export function deleteModelWithClass<U extends AnyParamConstructor<any>>(cl: U) {
  assertionIsClass(cl);
  assertion(isGlobalCachingEnabled(), () => new CacheDisabledError('deleteModelWithClass'));

  let name = getName(cl);

  if (!models.has(name)) {
    logger.debug(`Class "${name}" is not in "models", trying to find in "constructors"`);
    let found = false;

    // type "Map" does not have a "find" function, and using "get" would maybe result in the incorrect values
    for (const [cname, constructor] of constructors) {
      if (constructor === cl) {
        logger.debug(`Found Class in "constructors" with class name "${name}" and entered name "${cname}""`);
        name = cname;
        found = true;
      }
    }

    if (!found) {
      logger.debug(`Could not find class "${name}" in constructors`);

      return;
    }
  }

  return deleteModel(name);
}

/**
 * Build a Model from the given Class and add it as a discriminator onto "from"
 * @param from The Model to add the new discriminator model to
 * @param cl The Class to make a discriminator model from
 * @param options Overwrite ModelOptions (Merged with ModelOptions from class)
 * @example
 * ```ts
 * class Main {
 *   @prop({ ref: () => BaseDiscriminator })
 *   public discriminators?: Ref<BaseDiscriminator>;
 * }
 *
 * class BaseDiscriminator {
 *   @prop()
 *   public propertyOnAllDiscriminators?: string;
 * }
 *
 * class AnotherDiscriminator {
 *   @prop()
 *   public someValue?: string;
 * }
 *
 * const MainModel = getModelForClass(Main);
 *
 * const BaseDiscriminatorModel = getModelFroClass(BaseDiscriminator);
 * const AnotherDiscriminatorModel = getDiscriminatorModelForClass(BaseDiscriminatorModel, AnotherDiscriminator);
 * // add other discriminator models the same way as "AnotherDiscriminatorModel"
 * ```
 */
export function getDiscriminatorModelForClass<U extends AnyParamConstructor<any>, QueryHelpers = BeAnObject>(
  from: mongoose.Model<any, any, any, any>,
  cl: U,
  options?: IModelOptions
): ReturnModelType<U, QueryHelpers>;
/**
 * Build a Model from the given Class and add it as a discriminator onto "from"
 * @param from The Model to add the new discriminator model to
 * @param cl The Class to make a discriminator model from
 * @param value The Identifier to use to differentiate documents (default: cl.name)
 * @example
 * ```ts
 * class Main {
 *   @prop({ ref: () => BaseDiscriminator })
 *   public discriminators?: Ref<BaseDiscriminator>;
 * }
 *
 * class BaseDiscriminator {
 *   @prop()
 *   public propertyOnAllDiscriminators?: string;
 * }
 *
 * class AnotherDiscriminator {
 *   @prop()
 *   public someValue?: string;
 * }
 *
 * const MainModel = getModelForClass(Main);
 *
 * const BaseDiscriminatorModel = getModelFroClass(BaseDiscriminator);
 * const AnotherDiscriminatorModel = getDiscriminatorModelForClass(BaseDiscriminatorModel, AnotherDiscriminator);
 * // add other discriminator models the same way as "AnotherDiscriminatorModel"
 * ```
 */
export function getDiscriminatorModelForClass<U extends AnyParamConstructor<any>, QueryHelpers = BeAnObject>(
  from: mongoose.Model<any, any, any, any>,
  cl: U,
  value?: string
): ReturnModelType<U, QueryHelpers>;
/**
 * Build a Model from the given Class and add it as a discriminator onto "from"
 * @param from The Model to add the new discriminator model to
 * @param cl The Class to make a discriminator model from
 * @param value The Identifier to use to differentiate documents (default: cl.name)
 * @param options Overwrite ModelOptions (Merged with ModelOptions from class)
 * @example
 * ```ts
 * class Main {
 *   @prop({ ref: () => BaseDiscriminator })
 *   public discriminators?: Ref<BaseDiscriminator>;
 * }
 *
 * class BaseDiscriminator {
 *   @prop()
 *   public propertyOnAllDiscriminators?: string;
 * }
 *
 * class AnotherDiscriminator {
 *   @prop()
 *   public someValue?: string;
 * }
 *
 * const MainModel = getModelForClass(Main);
 *
 * const BaseDiscriminatorModel = getModelFroClass(BaseDiscriminator);
 * const AnotherDiscriminatorModel = getDiscriminatorModelForClass(BaseDiscriminatorModel, AnotherDiscriminator);
 * // add other discriminator models the same way as "AnotherDiscriminatorModel"
 * ```
 */
export function getDiscriminatorModelForClass<U extends AnyParamConstructor<any>, QueryHelpers = BeAnObject>(
  from: mongoose.Model<any, any, any, any>,
  cl: U,
  value?: string,
  options?: IModelOptions
): ReturnModelType<U, QueryHelpers>;
export function getDiscriminatorModelForClass<U extends AnyParamConstructor<any>, QueryHelpers = BeAnObject>(
  from: mongoose.Model<any, any, any, any>,
  cl: U,
  value_or_options?: string | IModelOptions,
  options?: IModelOptions
) {
  assertion(isModel(from), new NotValidModelError(from, 'getDiscriminatorModelForClass.from'));
  assertionIsClass(cl);

  const value = typeof value_or_options === 'string' ? value_or_options : undefined;
  const rawOptions = typeof value_or_options !== 'string' ? value_or_options : typeof options === 'object' ? options : {};
  const overwriteNaming = mapModelOptionsToNaming(rawOptions); // use "rawOptions" instead of "mergedOptions" to consistently differentiate between classes & models
  const mergedOptions = getMergedModelOptions(rawOptions, cl);
  const name = getName(cl, overwriteNaming);

  if (isCachingEnabled(mergedOptions.options?.disableCaching) && models.has(name)) {
    return models.get(name) as ReturnModelType<U, QueryHelpers>;
  }

  if (mergedOptions.existingConnection && mergedOptions.existingConnection !== from.db) {
    warnNotMatchingExisting(from.modelName, getName(cl), 'existingConnection');
  }
  if (mergedOptions.existingMongoose && mergedOptions.existingMongoose !== from.base) {
    warnNotMatchingExisting(from.modelName, getName(cl), 'existingMongoose');
  }

  const sch: mongoose.Schema<any> = buildSchema(cl, mergedOptions);

  const mergeHooks = mergedOptions.options?.enableMergeHooks ?? false;
  // Note: this option is not actually for "merging plugins", but if "true" it will *overwrite* all plugins with the base-schema's
  const mergePlugins = mergedOptions.options?.enableMergePlugins ?? false;

  const discriminatorKey = sch.get('discriminatorKey');

  if (!!discriminatorKey && sch.path(discriminatorKey)) {
    (sch.paths[discriminatorKey] as any).options.$skipDiscriminatorCheck = true;
  }

  const compiledModel = from.discriminator(name, sch, {
    value: value ? value : name,
    mergeHooks,
    mergePlugins,
  });

  return addModelToTypegoose<U, QueryHelpers>(compiledModel, cl, {
    disableCaching: mergedOptions.options?.disableCaching,
  });
}

/**
 * Use this class if raw mongoose for a path is wanted
 * It is still recommended to use the typegoose classes directly
 * @see Using `Passthrough`, the paths created will also result as an `Schema` (since mongoose 6.0), see {@link https://github.com/Automattic/mongoose/issues/7181 Mongoose#7181}
 * @example
 * ```ts
 * class Dummy {
 *   @prop({ type: () => new Passthrough({ somePath: String }) })
 *   public somepath: { somePath: string };
 * }
 *
 * class Dummy {
 *   @prop({ type: () => new Passthrough({ somePath: String }, true) })
 *   public somepath: { somePath: string };
 * }
 * ```
 */
export class Passthrough {
  // this property has no types, because it can slightly differentiate than a normal mongoose schema (like being a direct array)
  public raw: any;
  public direct: boolean;

  /**
   * Use this like `new mongoose.Schema()`
   * @param raw The Schema definition
   * @param direct Directly insert "raw", instead of using "type" (this will not apply any other inner options)
   */
  constructor(raw: any, direct?: boolean) {
    this.raw = raw;
    this.direct = direct ?? false;
  }
}
