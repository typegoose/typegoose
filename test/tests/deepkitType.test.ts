import { t, jsonSerializer } from '@deepkit/type';
import { DocumentType, getModelForClass, prop, mongoose } from '../../src/typegoose';

const ObjectId = mongoose.Types.ObjectId;
// Create a Custom Serializer to add custom transform functions to types
const mySerializer = new (class CustomSerializer extends jsonSerializer.fork('mySerializer') {})();

// Note: A custom Serializer has to be used, because the included "mongoId" "decorator" only works with "@deepkit/orm"

// We overwrite mongoId and correctly convert from Mongo ObjectID to string when deserializing
mySerializer.toClass.register('objectId', (property, state) => {
  state.setContext({ ObjectId: ObjectId });
  state.addSetter(`${state.accessor} instanceof String ? ObjectId.createFromHexString(${state.accessor}) : ${state.accessor}`);
});

// We overwrite mongoId and correctly convert string to Mongo ObjectID when serializing
mySerializer.fromClass.register('objectId', (property, state) => {
  state.setContext({ ObjectId: ObjectId });
  state.addSetter(`${state.accessor} instanceof ObjectId ? ${state.accessor}.toHexString() : ${state.accessor}`);
});

// Create a custom "classToPlain" function, using "mySerializer" instead of the function provided by "@deepkit/type"
const classToPlain = function (schemaCls: any, clsObj: any, access?: any) {
  return mySerializer.for(schemaCls).serialize(clsObj, access);
};

// Create a custom "plainToClass" function, using "mySerializer" instead of the function provided by "@deepkit/type"
const plainToClass = function (schemaCls: any, obj: any, access?: any) {
  return mySerializer.for(schemaCls).deserialize(obj, access);
};

enum Group {
  confidential = 'confidential',
  public = 'public',
}

class Account {
  @t.mongoId.group(Group.public)
  public _id: string;

  @t.group(Group.public)
  public __v: number;

  @t.group(Group.public)
  @prop()
  public email: string;

  @t.group(Group.confidential)
  @prop()
  public confidentialProperty?: string;
}

const AccountModel = getModelForClass(Account);

describe('@deepkit/type transforms', () => {
  let id: string;
  let accountClassObject: Account;
  let account: DocumentType<Account>;
  beforeEach(async () => {
    accountClassObject = new Account();
    accountClassObject.email = 'somebody@gmail.com';
    accountClassObject.confidentialProperty = 'secret';
    account = await AccountModel.create(accountClassObject);
    accountClassObject._id = account._id;
    accountClassObject.__v = account.__v;
    id = account._id.toString();
  });

  describe('lean query', () => {
    it('should be able to use @deepkit/type to transform to class instance', async () => {
      // lean return a Plain Old Javascript Object
      const pojo = await AccountModel.findById(id).orFail().lean().exec();
      // groupsExclude option excludes the property group
      const access = { groupsExclude: [Group.confidential] };
      // deserialize pojo back to an Account instance
      const deserialized = plainToClass(Account, pojo, access);
      // need to remove the password property, as it is not available due to our @t.group confidential exclusion definition
      delete accountClassObject.confidentialProperty;
      expect(deserialized).toStrictEqual(accountClassObject);
      expect(deserialized).not.toHaveProperty('confidentialProperty');
    });

    it("should be able to use @deepkit/type's groups to allow access", async () => {
      // lean return a Plain Old Javascript Object
      const pojo = await AccountModel.findById(id).orFail().lean().exec();
      // groups option includes the property group
      const access = { groups: [Group.confidential, Group.public] };
      // deserialize pojo back to an Account instance
      const deserialized = plainToClass(Account, pojo, access);
      expect(deserialized).toStrictEqual(accountClassObject);
      expect(deserialized).toHaveProperty('confidentialProperty');
    });
  });

  describe('normal query', () => {
    it('should be able to use @deepkit/type transform', async () => {
      // exec returns a Mongoose Object
      const doc = await AccountModel.findById(id).orFail().exec();
      const access = { groupsExclude: [Group.confidential] };
      // serialize Account instance back to a Plain Old Javascript Object
      const serialized = classToPlain(Account, doc, access);
      expect(serialized).toStrictEqual({
        _id: id,
        __v: 0,
        email: 'somebody@gmail.com',
      });
      expect(serialized).not.toHaveProperty('confidentialProperty');
    });

    it("should be able to use @deepkit/type's groups to allow access", async () => {
      // exec return a Mongoose Object
      const doc = await AccountModel.findById(id).orFail().exec();
      const access = { groups: [Group.confidential, Group.public] };
      // serialize Account instance back to a Plain Old Javascript Object
      const serialized = classToPlain(Account, doc, access);
      expect(serialized).toStrictEqual({
        _id: id,
        __v: 0,
        email: 'somebody@gmail.com',
        confidentialProperty: 'secret',
      });
      expect(serialized).toHaveProperty('confidentialProperty');
    });
  });

  it('should create a Document from a Class', async () => {
    const origData = new Account(); // data without any mongoose properties (like "__v" and "_id")
    origData.confidentialProperty = 'confident1';
    origData.email = 'nobody1@someone.org';
    // the following is to have a copy of the original, just to test "@deepkit/type"'s ability to transform
    const copied = plainToClass(Account, classToPlain(Account, origData));
    // this is here, because in the transition from mongoose 5.x to mongoose 6.x, class-transformer suddenly started having different values after one transform
    expect(copied).toStrictEqual(origData);
    const createdDoc = await AccountModel.create(copied);

    // Expect "createdDoc" to have the properties of "origData" exactly matching, ignoring extra properties that were not in "origData"
    expect(createdDoc).toStrictEqual(expect.objectContaining(origData));
    const docPOJO = classToPlain(Account, createdDoc);
    expect(docPOJO).toMatchSnapshot({ _id: expect.any(String) });
  });

  it('should transform a POJO to a Class, store it, retrieve it, and transform it back to a POJO', async () => {
    const origData = new Account(); // data without any mongoose properties (like "__v" and "_id")
    origData.confidentialProperty = 'confident2';
    origData.email = 'nobody2@someone.org';
    // create our definition for groupings of properties to transform
    const access = { groups: [Group.confidential, Group.public] };
    // transform our "fake" incoming DTO to a Class instance
    const incomingDataTransformedToClass = plainToClass(Account, origData, access);
    // Store the class object
    const createdDoc = await AccountModel.create(incomingDataTransformedToClass);
    // retrieve the document
    const docFound = await AccountModel.findById(createdDoc._id).orFail().exec();
    // transform it back to a DTO
    const outboundDTO = classToPlain(Account, docFound, access);
    expect(outboundDTO).toMatchSnapshot({ _id: expect.any(String) });
  });
});
