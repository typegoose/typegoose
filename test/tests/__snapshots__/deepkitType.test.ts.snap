// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@deepkit/type transforms should create a Document from a Class 1`] = `
Object {
  "__v": 0,
  "_id": Any<String>,
  "confidentialProperty": "confident1",
  "email": "nobody1@someone.org",
}
`;

exports[`@deepkit/type transforms should transform a POJO to a Class, store it, retrieve it, and transform it back to a POJO 1`] = `
Object {
  "__v": 0,
  "_id": Any<String>,
  "confidentialProperty": "confident2",
  "email": "nobody2@someone.org",
}
`;
